File: scripts/build-founding-promotion-review.mjs
import fs from 'node:fs';
import path from 'node:path';

const root = process.cwd();
const publicFoundingPath = path.join(root, 'public', 'company_founding_dates.csv');
const outPath = path.join(root, 'research', 'revisit-lists', 'founding-promotion-review.csv');
const DAY_MS = 24 * 60 * 60 * 1000;

const SOURCES = [
  { dir: path.join(root, 'research', 'founding-sec-batches'), pattern: /^founding-date-sec-candidates-\d+\.csv$/ },
  { dir: path.join(root, 'research', 'founding-batches'), pattern: /^founding-date-candidates-\d+\.csv$/ },
  { dir: path.join(root, 'research'), pattern: /^founding-date-site-candidates-\d+\.csv$/ },
];

const HEADERS = [
  'company',
  'latest_exit_date',
  'deal_type',
  'hq_location',
  'founding_date',
  'date_basis',
  'source',
  'confidence',
  'status',
  'time_to_exit_years',
  'recommendation',
  'review_reason',
  'matched_name',
  'url',
  'notes',
];

const csvEscape = (value) => {
  const text = String(value ?? '');
  return /[",\n]/.test(text) ? `"${text.replace(/"/g, '""')}"` : text;
};

const parseCsv = (text) => {
  const rows = [];
  let row = [];
  let cell = '';
  let inQuotes = false;

  for (let i = 0; i < text.length; i += 1) {
    const ch = text[i];
    const next = text[i + 1];
    if (ch === '"') {
      if (inQuotes && next === '"') {
        cell += '"';
        i += 1;
      } else {
        inQuotes = !inQuotes;
      }
    } else if (ch === ',' && !inQuotes) {
      row.push(cell);
      cell = '';
    } else if ((ch === '\n' || ch === '\r') && !inQuotes) {
      if (ch === '\r' && next === '\n') i += 1;
      row.push(cell);
      if (row.some((value) => value.trim())) rows.push(row);
      row = [];
      cell = '';
    } else {
      cell += ch;
    }
  }

  if (cell.length || row.length) {
    row.push(cell);
    if (row.some((value) => value.trim())) rows.push(row);
  }

  return rows;
};

const rowsFromCsv = (file) => {
  const rows = parseCsv(fs.readFileSync(file, 'utf8'));
  const headers = rows[0].map((header) => header.trim());
  return rows.slice(1).map((cols) => Object.fromEntries(headers.map((header, i) => [header, cols[i] ?? ''])));
};

const companyKey = (value) => String(value || '').trim().toLowerCase();

const parseDate = (value) => {
  if (!value) return null;
  const text = String(value).trim();
  let match = text.match(/^(\d{4})-(\d{2})-(\d{2})$/);
  if (match) return new Date(Date.UTC(+match[1], +match[2] - 1, +match[3]));
  match = text.match(/^(\d{4})-(\d{2})$/);
  if (match) return new Date(Date.UTC(+match[1], +match[2] - 1, 1));
  match = text.match(/^(\d{4})$/);
  if (match) return new Date(Date.UTC(+match[1], 0, 1));
  match = text.match(/(\d{1,2})-(\w{3})-(\d{4})/);
  if (match) return new Date(`${match[1]} ${match[2]} ${match[3]} UTC`);
  return null;
};

const yearsBetween = (startValue, endValue) => {
  const start = parseDate(startValue);
  const end = parseDate(endValue);
  if (!start || !end || end < start) return '';
  return (((end - start) / DAY_MS) / 365.25).toFixed(2);
};

const normalizeCandidateDate = (value) => {
  const text = String(value || '').trim();
  let match = text.match(/^(\d{4})-00-00$/);
  if (match) return match[1];
  match = text.match(/^(\d{4})-(\d{2})-00$/);
  if (match) return match[2] === '00' ? match[1] : `${match[1]}-${match[2]}`;
  match = text.match(/^(\d{4})-00$/);
  if (match) return match[1];
  return text;
};

const inferDateBasis = (row) => {
  if (row.date_basis) return row.date_basis;
  const text = `${row.source || ''} ${row.notes || ''}`.toLowerCase();
  if (/founded|inception|official site|wikidata/.test(text)) return 'operating_founding';
  if (/incorporated|incorporation/.test(text)) return 'legal_incorporation';
  return '';
};

const recommendationFor = (row, timeToExit) => {
  const basis = row.date_basis;
  const confidence = row.confidence || '';
  const status = row.status || '';
  const years = timeToExit === '' ? null : Number(timeToExit);
  const text = `${row.company || ''} ${row.matched_name || ''} ${row.notes || ''}`.toLowerCase();

  if (!row.founding_date) return ['SKIP', 'No founding date candidate'];
  if (!['SEC_MATCH', 'LIKELY_MATCH', 'SITE_MATCH', 'REVIEW'].includes(status)) return ['SKIP', `Candidate status is ${status || 'blank'}`];
  if (/spac|blank check|special purpose acquisition|successor|reorganization|holding company/.test(text)) {
    return ['REVIEW', 'Possible issuer, SPAC, holding-company, or reorganization date'];
  }
  if (basis === 'operating_founding' && ['HIGH', 'MEDIUM_HIGH', 'MEDIUM'].includes(confidence)) {
    if (row.source === 'Wikidata' && confidence !== 'HIGH') {
      return ['REVIEW', 'Wikidata match below HIGH confidence needs review'];
    }
    return ['PROMOTE_CANDIDATE', 'Operating founding date with usable confidence'];
  }
  if (basis === 'legal_incorporation' && years !== null && years < 3) {
    return ['REVIEW', 'Very short legal age may be issuer/reorg incorporation'];
  }
  if (basis === 'legal_incorporation') {
    return ['REVIEW', 'Legal incorporation needs operating-company review'];
  }
  return ['REVIEW', 'Needs manual basis review'];
};

const existingRows = fs.existsSync(publicFoundingPath) ? rowsFromCsv(publicFoundingPath) : [];
const existingCompanies = new Set(existingRows.map((row) => companyKey(row.company)));
const seen = new Set();
const candidates = [];

for (const source of SOURCES) {
  if (!fs.existsSync(source.dir)) continue;
  for (const file of fs.readdirSync(source.dir).filter((name) => source.pattern.test(name)).sort()) {
    for (const row of rowsFromCsv(path.join(source.dir, file))) {
      if (!row.company || existingCompanies.has(companyKey(row.company)) || !row.founding_date) continue;
      const key = `${companyKey(row.company)}|${row.founding_date}|${row.source || ''}`;
      if (seen.has(key)) continue;
      seen.add(key);
      const enriched = { ...row, founding_date: normalizeCandidateDate(row.founding_date), date_basis: inferDateBasis(row) };
      const timeToExit = yearsBetween(enriched.founding_date, enriched.latest_exit_date);
      const [recommendation, reviewReason] = recommendationFor(enriched, timeToExit);
      candidates.push({
        ...enriched,
        time_to_exit_years: timeToExit,
        recommendation,
        review_reason: reviewReason,
        url: enriched.filing_url || enriched.page_url || '',
      });
    }
  }
}

candidates.sort((a, b) => {
  const rank = { PROMOTE_CANDIDATE: 0, REVIEW: 1, SKIP: 2 };
  return (rank[a.recommendation] ?? 9) - (rank[b.recommendation] ?? 9)
    || String(a.company).localeCompare(String(b.company));
});

fs.mkdirSync(path.dirname(outPath), { recursive: true });
fs.writeFileSync(
  outPath,
  [HEADERS, ...candidates.map((row) => HEADERS.map((header) => row[header] ?? ''))]
    .map((row) => row.map(csvEscape).join(','))
    .join('\n'),
);

const summary = candidates.reduce((acc, row) => {
  acc[row.recommendation] = (acc[row.recommendation] || 0) + 1;
  return acc;
}, {});

console.log(JSON.stringify({ output: outPath, rows: candidates.length, summary }, null, 2));

File: scripts/build_founding_promotion_review.py
#!/usr/bin/env python3
"""Python port of scripts/build-founding-promotion-review.mjs (no Node available in this env)."""
import csv
import glob
import json
import os
import re

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
PUBLIC_FOUNDING_PATH = os.path.join(ROOT, 'public', 'company_founding_dates.csv')
OUT_PATH = os.path.join(ROOT, 'research', 'revisit-lists', 'founding-promotion-review.csv')

SOURCES = [
    (os.path.join(ROOT, 'research', 'founding-sec-batches'), re.compile(r'^founding-date-sec-candidates-\d+\.csv$')),
    (os.path.join(ROOT, 'research', 'founding-batches'), re.compile(r'^founding-date-candidates-\d+\.csv$')),
    (os.path.join(ROOT, 'research'), re.compile(r'^founding-date-site-candidates-\d+\.csv$')),
]

HEADERS = [
    'company', 'latest_exit_date', 'deal_type', 'hq_location', 'founding_date', 'date_basis',
    'source', 'confidence', 'status', 'time_to_exit_years', 'recommendation', 'review_reason',
    'matched_name', 'url', 'notes',
]


def csv_escape(value):
    text = '' if value is None else str(value)
    if re.search(r'[",\n]', text):
        return '"' + text.replace('"', '""') + '"'
    return text


def company_key(value):
    return str(value or '').strip().lower()


MONTH_ABBR = {m: i + 1 for i, m in enumerate(['Jan','Feb','Mar','Apr','May','Jun','Jul','Aug','Sep','Oct','Nov','Dec'])}


def parse_date(value):
    if not value:
        return None
    text = str(value).strip()
    m = re.match(r'^(\d{4})-(\d{2})-(\d{2})$', text)
    if m:
        import datetime
        return datetime.date(int(m[1]), int(m[2]), int(m[3]))
    m = re.match(r'^(\d{4})-(\d{2})$', text)
    if m:
        import datetime
        return datetime.date(int(m[1]), int(m[2]), 1)
    m = re.match(r'^(\d{4})$', text)
    if m:
        import datetime
        return datetime.date(int(m[1]), 1, 1)
    m = re.match(r'^(\d{1,2})-(\w{3})-(\d{4})', text)
    if m and m[2] in MONTH_ABBR:
        import datetime
        return datetime.date(int(m[3]), MONTH_ABBR[m[2]], int(m[1]))
    return None


def years_between(start_value, end_value):
    start = parse_date(start_value)
    end = parse_date(end_value)
    if not start or not end or end < start:
        return ''
    return f'{(end - start).days / 365.25:.2f}'


def normalize_candidate_date(value):
    text = str(value or '').strip()
    m = re.match(r'^(\d{4})-00-00$', text)
    if m:
        return m[1]
    m = re.match(r'^(\d{4})-(\d{2})-00$', text)
    if m:
        return m[1] if m[2] == '00' else f'{m[1]}-{m[2]}'
    m = re.match(r'^(\d{4})-00$', text)
    if m:
        return m[1]
    return text


def infer_date_basis(row):
    if row.get('date_basis'):
        return row['date_basis']
    text = f"{row.get('source', '')} {row.get('notes', '')}".lower()
    if re.search(r'founded|inception|official site|wikidata', text):
        return 'operating_founding'
    if re.search(r'incorporated|incorporation', text):
        return 'legal_incorporation'
    return ''


def recommendation_for(row, time_to_exit):
    basis = row.get('date_basis')
    confidence = row.get('confidence') or ''
    status = row.get('status') or ''
    years = None if time_to_exit == '' else float(time_to_exit)
    text = f"{row.get('company', '')} {row.get('matched_name', '')} {row.get('notes', '')}".lower()

    if not row.get('founding_date'):
        return 'SKIP', 'No founding date candidate'
    if status not in ('SEC_MATCH', 'LIKELY_MATCH', 'SITE_MATCH', 'REVIEW'):
        return 'SKIP', f'Candidate status is {status or "blank"}'
    if re.search(r'spac|blank check|special purpose acquisition|successor|reorganization|holding company', text):
        return 'REVIEW', 'Possible issuer, SPAC, holding-company, or reorganization date'
    if basis == 'operating_founding' and confidence in ('HIGH', 'MEDIUM_HIGH', 'MEDIUM'):
        if row.get('source') == 'Wikidata' and confidence != 'HIGH':
            return 'REVIEW', 'Wikidata match below HIGH confidence needs review'
        return 'PROMOTE_CANDIDATE', 'Operating founding date with usable confidence'
    if basis == 'legal_incorporation' and years is not None and years < 3:
        return 'REVIEW', 'Very short legal age may be issuer/reorg incorporation'
    if basis == 'legal_incorporation':
        return 'REVIEW', 'Legal incorporation needs operating-company review'
    return 'REVIEW', 'Needs manual basis review'


def rows_from_csv(path):
    with open(path, newline='', encoding='utf-8') as f:
        return list(csv.DictReader(f))


def main():
    existing_rows = rows_from_csv(PUBLIC_FOUNDING_PATH) if os.path.exists(PUBLIC_FOUNDING_PATH) else []
    existing_companies = {company_key(r['company']) for r in existing_rows}
    seen = set()
    candidates = []

    for source_dir, pattern in SOURCES:
        if not os.path.isdir(source_dir):
            continue
        for name in sorted(os.listdir(source_dir)):
            if not pattern.match(name):
                continue
            for row in rows_from_csv(os.path.join(source_dir, name)):
                if not row.get('company') or company_key(row['company']) in existing_companies or not row.get('founding_date'):
                    continue
                key = f"{company_key(row['company'])}|{row['founding_date']}|{row.get('source', '')}"
                if key in seen:
                    continue
                seen.add(key)
                enriched = {**row, 'founding_date': normalize_candidate_date(row['founding_date'])}
                enriched['date_basis'] = infer_date_basis(enriched)
                time_to_exit = years_between(enriched['founding_date'], enriched.get('latest_exit_date'))
                recommendation, review_reason = recommendation_for(enriched, time_to_exit)
                candidates.append({
                    **enriched,
                    'time_to_exit_years': time_to_exit,
                    'recommendation': recommendation,
                    'review_reason': review_reason,
                    'url': enriched.get('filing_url') or enriched.get('page_url') or '',
                })

    rank = {'PROMOTE_CANDIDATE': 0, 'REVIEW': 1, 'SKIP': 2}
    candidates.sort(key=lambda r: (rank.get(r['recommendation'], 9), str(r.get('company', ''))))

    os.makedirs(os.path.dirname(OUT_PATH), exist_ok=True)
    with open(OUT_PATH, 'w', encoding='utf-8') as f:
        f.write(','.join(HEADERS) + '\n')
        for row in candidates:
            f.write(','.join(csv_escape(row.get(h, '')) for h in HEADERS) + '\n')

    summary = {}
    for row in candidates:
        summary[row['recommendation']] = summary.get(row['recommendation'], 0) + 1

    print(json.dumps({'output': OUT_PATH, 'rows': len(candidates), 'summary': summary}, indent=2))


if __name__ == '__main__':
    main()
